File: relationship-manager/followup-orchestrator/src/main/java/com/bank/rm/followup/service/FollowUpService.java
package com.bank.rm.followup.service;

import com.bank.rm.common.dto.FollowUpFrequency;
import com.bank.rm.common.dto.FollowUpStatus;
import com.bank.rm.common.event.FollowUpDueEvent;
import com.bank.rm.common.event.FollowUpScheduledEvent;
import com.bank.rm.common.event.KafkaTopics;
import com.bank.rm.followup.domain.FollowUpInstance;
import com.bank.rm.followup.domain.FollowUpSchedule;
import com.bank.rm.followup.dto.FollowUpDtos.*;
import com.bank.rm.followup.repository.FollowUpInstanceRepository;
import com.bank.rm.followup.repository.FollowUpScheduleRepository;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.langchain4j.model.chat.ChatLanguageModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.Instant;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class FollowUpService {
    private static final Logger log = LoggerFactory.getLogger(FollowUpService.class);
    private final FollowUpScheduleRepository scheduleRepository;
    private final FollowUpInstanceRepository instanceRepository;
    private final ChatLanguageModel chatModel;
    private final KafkaTemplate<String, String> kafkaTemplate;
    private final ObjectMapper objectMapper;

    public FollowUpService(FollowUpScheduleRepository scheduleRepository,
                            FollowUpInstanceRepository instanceRepository,
                            ChatLanguageModel chatModel,
                            KafkaTemplate<String, String> kafkaTemplate,
                            ObjectMapper objectMapper) {
        this.scheduleRepository = scheduleRepository;
        this.instanceRepository = instanceRepository;
        this.chatModel = chatModel;
        this.kafkaTemplate = kafkaTemplate;
        this.objectMapper = objectMapper;
    }

    @Transactional
    public ScheduleResponse createSchedule(CreateScheduleRequest request) {
        UUID scheduleId = UUID.randomUUID();
        LocalDateTime nextFollowUp = computeNextFollowUp(request.frequency(), LocalDateTime.now());
        FollowUpSchedule schedule = new FollowUpSchedule(
                scheduleId, request.customerId(), request.frequency(), request.channel(), nextFollowUp);
        scheduleRepository.save(schedule);

        publishEvent(KafkaTopics.FOLLOWUP_EVENTS,
                new FollowUpScheduledEvent(request.customerId(), scheduleId, request.frequency(), request.channel()));

        return new ScheduleResponse(scheduleId, request.customerId(), request.frequency(),
                request.channel(), nextFollowUp, true, 0);
    }

    @Scheduled(fixedRate = 60000)
    @Transactional
    public void processScheduledFollowUps() {
        List<FollowUpSchedule> due = scheduleRepository.findDueSchedules(LocalDateTime.now());
        for (FollowUpSchedule schedule : due) {
            try {
                String agenda = generateAgenda(schedule.getCustomerId());
                FollowUpInstance instance = new FollowUpInstance(
                        UUID.randomUUID(), schedule.getId(), schedule.getCustomerId(),
                        schedule.getNextFollowUpAt(), agenda);
                instanceRepository.save(instance);

                publishEvent(KafkaTopics.FOLLOWUP_EVENTS,
                        new FollowUpDueEvent(schedule.getCustomerId(), instance.getId(), schedule.getChannel()));

                schedule.setNextFollowUpAt(computeNextFollowUp(schedule.getFrequency(), schedule.getNextFollowUpAt()));
                schedule.setUpdatedAt(Instant.now());
                scheduleRepository.save(schedule);
            } catch (Exception e) {
                log.error("Failed to process follow-up for customer {}: {}", schedule.getCustomerId(), e.getMessage());
            }
        }
    }

    private String generateAgenda(UUID customerId) {
        try {
            String prompt = "Generate a brief follow-up meeting agenda for a banking customer review. " +
                    "Include: portfolio performance review, risk assessment update, new product recommendations, " +
                    "and action items from previous meeting. Return as JSON array of agenda items.";
            return chatModel.generate(prompt);
        } catch (Exception e) {
            return "[\"Review portfolio performance\",\"Update risk profile\",\"Discuss new recommendations\",\"Review action items\"]";
        }
    }

    private LocalDateTime computeNextFollowUp(FollowUpFrequency frequency, LocalDateTime from) {
        return switch (frequency) {
            case WEEKLY -> from.plusWeeks(1);
            case BIWEEKLY -> from.plusWeeks(2);
            case MONTHLY -> from.plusMonths(1);
            case QUARTERLY -> from.plusMonths(3);
        };
    }

    private void publishEvent(String topic, Object event) {
        try {
            kafkaTemplate.send(topic, objectMapper.writeValueAsString(event));
        } catch (Exception e) { log.error("Failed to publish event: {}", e.getMessage()); }
    }
}

File: relationship-manager/followup-orchestrator/src/main/kotlin/com/bank/rm/followup/service/FollowUpService.kt
package com.bank.rm.followup.service

import com.bank.rm.common.dto.*
import com.bank.rm.common.event.*
import com.bank.rm.common.exception.ResourceNotFoundException
import com.bank.rm.followup.domain.*
import com.bank.rm.followup.dto.*
import com.bank.rm.followup.repository.*
import dev.langchain4j.model.chat.ChatLanguageModel
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import java.util.UUID

@Service
class FollowUpService(
    private val scheduleRepository: FollowUpScheduleRepository,
    private val instanceRepository: FollowUpInstanceRepository,
    private val actionItemRepository: ActionItemRepository,
    private val chatModel: ChatLanguageModel,
    private val kafkaTemplate: KafkaTemplate<String, DomainEvent>
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Transactional
    fun createSchedule(customerId: UUID, request: CreateScheduleRequest): ScheduleResponse {
        val nextFollowUp = calculateNextFollowUp(request.frequency, request.startDate)

        val schedule = FollowUpSchedule(
            customerId = customerId,
            frequency = request.frequency,
            channel = request.channel,
            nextFollowUpAt = nextFollowUp
        )
        scheduleRepository.save(schedule)

        // Create first instance
        val instance = FollowUpInstance(
            scheduleId = schedule.id,
            customerId = customerId,
            scheduledAt = nextFollowUp,
            agenda = generateAgenda(customerId, null)
        )
        instanceRepository.save(instance)

        kafkaTemplate.send(
            KafkaTopics.FOLLOWUP_EVENTS,
            customerId.toString(),
            FollowUpScheduledEvent(
                customerId = customerId,
                scheduleId = schedule.id,
                frequency = request.frequency,
                channel = request.channel
            )
        )

        return schedule.toResponse()
    }

    fun getSchedule(customerId: UUID): ScheduleResponse? {
        val schedule = scheduleRepository.findByCustomerIdAndIsActiveTrue(customerId) ?: return null
        return schedule.toResponse()
    }

    fun getUpcomingInstances(customerId: UUID): List<FollowUpInstanceResponse> =
        instanceRepository.findByCustomerIdAndStatusOrderByScheduledAtAsc(customerId, FollowUpStatus.SCHEDULED)
            .map { it.toResponse() }

    @Transactional
    fun completeFollowUp(instanceId: UUID, conversationId: UUID?): FollowUpInstanceResponse {
        val instance = instanceRepository.findById(instanceId).orElseThrow {
            ResourceNotFoundException("Follow-up instance not found: $instanceId")
        }

        instance.status = FollowUpStatus.COMPLETED
        instance.completedAt = Instant.now()
        instance.conversationId = conversationId
        instanceRepository.save(instance)

        // Update schedule
        val schedule = scheduleRepository.findById(instance.scheduleId).orElse(null)
        if (schedule != null) {
            schedule.totalCompleted++
            val nextFollowUp = calculateNextFollowUp(schedule.frequency, schedule.nextFollowUpAt)
            schedule.nextFollowUpAt = nextFollowUp
            schedule.updatedAt = Instant.now()
            scheduleRepository.save(schedule)

            // Create next instance
            val nextInstance = FollowUpInstance(
                scheduleId = schedule.id,
                customerId = schedule.customerId,
                scheduledAt = nextFollowUp,
                agenda = generateAgenda(schedule.customerId, instance.id)
            )
            instanceRepository.save(nextInstance)
        }

        return instance.toResponse()
    }

    // Processes due follow-ups every hour
    @Scheduled(fixedRate = 3600000)
    fun processDueFollowUps() {
        val now = LocalDateTime.now()
        val dueInstances = instanceRepository.findDueInstances(now)

        dueInstances.forEach { instance ->
            if (instance.reminderSentAt == null) {
                val schedule = scheduleRepository.findById(instance.scheduleId).orElse(null)
                if (schedule != null) {
                    kafkaTemplate.send(
                        KafkaTopics.REMINDER_EVENTS,
                        instance.customerId.toString(),
                        FollowUpDueEvent(
                            customerId = instance.customerId,
                            instanceId = instance.id,
                            channel = schedule.channel
                        )
                    )
                    instance.reminderSentAt = Instant.now()
                    instance.status = FollowUpStatus.REMINDER_SENT
                    instanceRepository.save(instance)
                }
            }
        }
    }

    @KafkaListener(topics = [KafkaTopics.CONVERSATION_EVENTS], groupId = "followup-orchestrator")
    fun handleConversationEvent(event: ConversationCompletedEvent) {
        log.info("Conversation completed for customer {}: scheduling follow-ups", event.customerId)
    }

    private fun generateAgenda(customerId: UUID, previousInstanceId: UUID?): String {
        return try {
            val previousSummary = previousInstanceId?.let {
                instanceRepository.findById(it).orElse(null)?.agenda
            }

            val prompt = """Generate a brief follow-up agenda for a banking relationship review meeting.
${if (previousSummary != null) "Previous meeting notes: $previousSummary" else "This is the first follow-up meeting."}

Include 3-4 bullet points covering:
1. Review of current portfolio performance
2. Any market changes relevant to their risk profile
3. Action items from previous meeting (if any)
4. Topics for discussion

Return as JSON: {"items": ["item1", "item2", ...]}"""

            chatModel.generate(prompt)
        } catch (e: Exception) {
            log.warn("AI agenda generation failed: ${e.message}")
            """{"items": ["Review portfolio performance", "Discuss market updates", "Review action items", "Plan next steps"]}"""
        }
    }

    private fun calculateNextFollowUp(frequency: FollowUpFrequency, from: LocalDateTime?): LocalDateTime {
        val base = from ?: LocalDateTime.now()
        return when (frequency) {
            FollowUpFrequency.WEEKLY -> base.plusWeeks(1)
            FollowUpFrequency.BIWEEKLY -> base.plusWeeks(2)
            FollowUpFrequency.MONTHLY -> base.plusMonths(1)
            FollowUpFrequency.QUARTERLY -> base.plusMonths(3)
        }
    }

    private fun FollowUpSchedule.toResponse() = ScheduleResponse(
        id, customerId, frequency, channel, nextFollowUpAt.toString(), isActive, totalCompleted, totalMissed
    )

    private fun FollowUpInstance.toResponse() = FollowUpInstanceResponse(
        id, scheduleId, customerId, scheduledAt.toString(), status, agenda, conversationId
    )
}
